Add more integration configuration examples to the Ruby Heroku guide (#12338)
DataDog/documentation: content/en/agent/guide/heroku-ruby.md

## Setting up integrations

> Datadog comes with more than 400 turn-key integrations that collect metrics from different tech stacks. The Datadog buildpack allows you to enable these integrations for your Heroku application. We have added four examples to this guide for commonly used integrations in Heroku.

Reviewer: ```suggestion
Datadog comes with more than 400 turn-key integrations that collect metrics from different tech stacks. The Datadog buildpack allows you to enable these integrations for your Heroku application. 

Four commonly-used Heroku integration configuration examples are listed below.
```

After, you can successfully see the `/widgets` endpoint of your application, which uses that database.

> To enable the Postgres Datadog integration you need to retrieve the database credentials from Heroku:

Reviewer: ```suggestion
To enable the Postgres Datadog integration, retrieve the database credentials from Heroku:
```

Create a configuration file called `postgres.yaml` replacing with your host, dbname, username, and password with the information you got in the previous command:

```yaml
init_config:

instances:
  - host: <YOUR HOSTNAME>
    port: <YOUR PORT>
    username: <YOUR USERNAME>
    password: <YOUR PASSWORD>
    dbname: <YOUR DBNAME>
    ssl: True
```

> Instead of manually replacing and harcoding the configuration, you can also set up your Postgres integration based on Heroku environment variables, using the [prerun script][14] to replace those values before starting the Datadog agent:

Reviewer: ```suggestion
Instead of manually updating the configuration, you can set up your Postgres integration based on Heroku environment variables using the [prerun script][14] to replace those values before starting the Datadog Agent:
```

[...]

  postgres (5.4.0)
  ----------------
    Instance ID: postgres:e07ef94b907fe733 [OK]
    Configuration Source: file:/app/.apt/etc/datadog-agent/conf.d/postgres.d/conf.yaml
    Total Runs: 9
    Metric Samples: Last Run: 15, Total: 135
    Events: Last Run: 0, Total: 0
    Service Checks: Last Run: 1, Total: 9
    Average Execution Time : 102ms
    Last Execution Date : 2021-05-11 14:14:34 UTC (1620742474000)
    Last Successful Execution Date : 2021-05-11 14:14:34 UTC (1620742474000)
    metadata:
      version.major: 13
      version.minor: 2
      version.patch: 0
      version.raw: 13.2 (Ubuntu 13.2-1.pgdg20.04+1)
      version.scheme: semver

[...]
```

Once you have checked that the Postgres check is running correctly, you can start looking at the Postgres metrics available on the [Metrics Summary][15] page:

{{< img src="agent/guide/heroku_ruby/postgres_metrics.png" alt="Datadog Metrics Explorer" >}}

### Redis

> For Redis, we will attach the [Heroku Redis addon][16] to our Heroku application:

Reviewer: ```suggestion
For Redis, attach the [Heroku Redis add on][16] to your Heroku application:
```

```shell
heroku addons:create heroku-redis:hobby-dev
```

> To check that Redis has been succesfully attached to your application, you can run the following command:

Reviewer: ```suggestion
To validate that Redis has successfully attached to your application, run the following command:
```

 ```shell
heroku addons:info REDIS
```

> You should get an output similar to the following:

Reviewer: The following output displays:

```shell
=== redis-cylindrical-59589
Attachments:  ruby-heroku-datadog::REDIS
Installed at: Wed Nov 17 2021 14:14:13 GMT+0100 (Central European Standard Time)
Owning app:   ruby-heroku-datadog
Plan:         heroku-redis:hobby-dev
Price:        free
State:        created
```

> And retrieve the credentials from HEROKU by running the following command:

Reviewer: ```suggestion
Retrieve the credentials from Heroku by running the following command:
```

```shell
heroku config -a $APPNAME | grep REDIS_URL
```

> Create a configuration file called `/datadog/conf.d/redisdb.yaml` at the root of your application replacing your host, port and password with the information you got in the previous command:

Reviewer: ```suggestion
Create a configuration file named `/datadog/conf.d/redisdb.yaml` at the root of your application to replace your host, port, and password with information from the previous command:
```

```yaml
init_config:

instances:
  - host: <YOUR_REDIS_HOST>
    password: <YOUR_REDIS_PASSWORD>
    port: <YOUR_REDIS_PORT>
```

> Instead of manually replacing and harcoding the configuration, you can also set up your Redis integration based on Heroku environment variables, using the [prerun script][14] to replace those values before starting the Datadog agent:

Reviewer: ```suggestion
Instead of manually updating the configuration, you can set up your Redis integration based on Heroku environment variables using the [prerun script][14] to replace those values before starting the Datadog Agent:
```

```shell
#!/usr/bin/env bash

# Update the Redis configuration from above using the Heroku application environment variable
if [ -n "$REDIS_URL" ]; then
  REDISREGEX='redis://([^:]*):([^@]+)@([^:]+):([^/]+)$'
  if [[ $REDIS_URL =~ $REDISREGEX ]]; then
    sed -i "s/<YOUR_REDIS_HOST>/${BASH_REMATCH[3]}/" "$DD_CONF_DIR/conf.d/redisdb.d/conf.yaml"
    sed -i "s/<YOUR_REDIS_PASSWORD>/${BASH_REMATCH[2]}/" "$DD_CONF_DIR/conf.d/redisdb.d/conf.yaml"
    sed -i "s/<YOUR_REDIS_PORT>/${BASH_REMATCH[4]}/" "$DD_CONF_DIR/conf.d/redisdb.d/conf.yaml"
  fi
fi
```

Deploy to Heroku:

```shell
# Deploy to Heroku
git add .
git commit -m "Enable redis integration"
git push heroku main
```

> Once the build finishes, the Datadog Agent starts the Redis check. Run the Datadog Agent status as explained in the [appendix section](#appendix-getting-the-datadog-agent-status) to make sure the Redis check is running correctly. You should look out for the following section:

Reviewer: ```suggestion
Once the build finishes, the Datadog Agent starts the Redis check. [Run the Datadog Agent status](#appendix-getting-the-datadog-agent-status) to make sure the Redis check is running correctly. 

The following output displays:
```

```

[...]

=========
Collector
=========

  Running Checks
  ==============

[...]

  redisdb (4.1.0)
  ---------------
    Instance ID: redisdb:eb3a3807075f89f0 [OK]
    Configuration Source: file:/app/.apt/etc/datadog-agent/conf.d/redisdb.d/conf.yaml
    Total Runs: 3
    Metric Samples: Last Run: 45, Total: 135
    Events: Last Run: 0, Total: 0
    Service Checks: Last Run: 1, Total: 3
    Average Execution Time : 6ms
    Last Execution Date : 2021-11-17 13:56:17 UTC (1637157377000)
    Last Successful Execution Date : 2021-11-17 13:56:17 UTC (1637157377000)
    metadata:
      version.major: 6
      version.minor: 2
      version.patch: 3
      version.raw: 6.2.3
      version.scheme: semver

[...]

```

### Sidekiq

> Sidekiq is a background processing framework for Ruby. If you are using Sidekiq Pro or Enterprise you can install the Datadog integration for Sidekiq following these instructions.

Reviewer: ```suggestion
Sidekiq is a background processing framework for Ruby. If you are using Sidekiq Pro or Enterprise, you can install the Datadog integration for Sidekiq.
```

> Install the `dogstatsd-ruby`:

Reviewer: ```suggestion
Install the `dogstatsd-ruby` package:
```

```shell
gem install dogstatsd-ruby
```

> Enable Sidekiq Pro metric collection by including this in your initializer:

Reviewer: ```suggestion
Enable Sidekiq Pro metric collection in your initializer:
```

```ruby
    require 'datadog/statsd' # gem 'dogstatsd-ruby'

    Sidekiq::Pro.dogstatsd = ->{ Datadog::Statsd.new('localhost', 8125, namespace:'sidekiq') }

    Sidekiq.configure_server do |config|
      config.server_middleware do |chain|
        require 'sidekiq/middleware/server/statsd'
        chain.add Sidekiq::Middleware::Server::Statsd
      end
    end
```

> If you are using Sidekiq Enterprise and would like to collect historical metrics, include this line as well:

Reviewer: ```suggestion
If you are using Sidekiq Enterprise and want to collect historical metrics, include the following:
```

```ruby
      Sidekiq.configure_server do |config|
        # history is captured every 30 seconds by default
        config.retain_history(30)
      end
```

Add the following to your [`datadog/prerun.sh`][14] script:

```
cat << 'EOF' >> "$DATADOG_CONF"

dogstatsd_mapper_profiles:
  - name: sidekiq
    prefix: "sidekiq."
    mappings:
      - match: 'sidekiq\.sidekiq\.(.*)'
        match_type: "regex"
        name: "sidekiq.$1"
      - match: 'sidekiq\.jobs\.(.*)\.perform'
        name: "sidekiq.jobs.perform"
        match_type: "regex"
        tags:
          worker: "$1"
      - match: 'sidekiq\.jobs\.(.*)\.(count|success|failure)'
        name: "sidekiq.jobs.worker.$2"
        match_type: "regex"
        tags:
          worker: "$1"
EOF
```

Deploy to Heroku:

```shell
# Deploy to Heroku
git add .
git commit -m "Enable sidekiq integration"
git push heroku main
```

> Once the build finishes, the Datadog Agent starts the Sidekiq check. Run the Datadog Agent status as explained in the [appendix section](#appendix-getting-the-datadog-agent-status) to make sure the Sidekiq check is running correctly.

Reviewer: ```suggestion
Once the build finishes, the Datadog Agent starts the Sidekiq check. [Run the Datadog Agent status](#appendix-getting-the-datadog-agent-status) to make sure the Sidekiq check is running correctly.
```

### Memcached

> Memcached is a distributed memory object caching system that is very popular in Rails applications. For this example, we will attach the [Heroku Memcached Cloud Addon][17] to our Heroku application:

Reviewer: ```suggestion
Memcached is a distributed memory object caching system that is popular in Rails applications. In this example, you can attach the [Heroku Memcached Cloud add on][17] to your Heroku application:
```

```
heroku addons:create memcachedcloud:30
```

> To check that Memcached has been succesfully attached to your application, you can run the following command:

Reviewer: ```suggestion
To check that Memcached has been successfully attached to your application, run the following command:
```

```shell
heroku addons | grep -A2 memcachedcloud
```

> You should get an output similar to this one:

Reviewer: ```suggestion
The following output displays:
```

```shell
memcachedcloud (memcachedcloud-fluffy-34783)   30         free   created
 └─ as MEMCACHEDCLOUD
```

> And retrieve the credentials from HEROKU by running the following command:

Reviewer: ```suggestion
Retrieve the credentials from Heroku by running the following:
```

```shell
heroku config | grep MEMCACHEDCLOUD
```

> Create a configuration file called `/datadog/conf.d/mcache.yaml` at the root of your application replacing your host, port, username and password with the information you got in the previous command:

Reviewer: ```suggestion
Create a configuration file named `/datadog/conf.d/mcache.yaml` at the root of your application to replace your host, port, username, and password with information from the previous command:
```

```yaml
instances:
  - url: <YOUR_MCACHE_HOST> 
    port: <YOUR_MCACHE_PORT>
    username: <YOUR_MCACHE_USERNAME>
    password: <YOUR_MCACHE_PASSWORD>
```

> Instead of manually replacing and harcoding the configuration, you can also set up your Memcached integration based on Heroku environment variables, using the [prerun script][14] to replace those values before starting the Datadog agent:

Reviewer: ```suggestion
Instead of manually updating the configuration, you can set up your Memcached integration based on Heroku environment variables using the [prerun script][14] to replace those values before starting the Datadog Agent:
```

```shell
#!/usr/bin/env bash

# Update the Memcached configuration from above using the Heroku application environment variable
if [ -n "$MEMCACHEDCLOUD_SERVERS" ]; then
  MCACHEREGEX='([^:]+):([^/]+)$'
  if [[ $MEMCACHEDCLOUD_SERVERS =~ $MCACHEREGEX ]]; then
    sed -i "s/<YOUR_MCACHE_HOST>/${BASH_REMATCH[1]}/" "$DD_CONF_DIR/conf.d/mcache.d/conf.yaml"
    sed -i "s/<YOUR_MCACHE_PORT>/${BASH_REMATCH[2]}/" "$DD_CONF_DIR/conf.d/mcache.d/conf.yaml"
  fi
  sed -i "s/<YOUR_MCACHE_USERNAME>/${MEMCACHEDCLOUD_USERNAME}/" "$DD_CONF_DIR/conf.d/mcache.d/conf.yaml"
  sed -i "s/<YOUR_MCACHE_PASSWORD>/${MEMCACHEDCLOUD_PASSWORD}/" "$DD_CONF_DIR/conf.d/mcache.d/conf.yaml"
fi
```

Deploy to Heroku:

```shell
# Deploy to Heroku
git add .
git commit -m "Enable memcached integration"
git push heroku main
```

> Once the build finishes, the Datadog Agent starts the Memcached check. Run the Datadog Agent status as explained in the [appendix section](#appendix-getting-the-datadog-agent-status) to make sure the Memcached check is running correctly. You should look out for the following section:

Reviewer: ```suggestion
Once the build finishes, the Datadog Agent starts the Memcached check. [Run the Datadog Agent status](#appendix-getting-the-datadog-agent-status) to make sure the Memcached check is running correctly. 

The following output displays:
```
